gh-126165: Improve docs of function `math.isclose` (#126215)
python/cpython: Doc/library/cmath.rst

   tolerance is ``1e-09``, which assures that the two values are the same
   within about 9 decimal digits.  *rel_tol* must be greater than zero.

>    *abs_tol* is the absolute tolerance; it must be at least ``0.0``.

Reviewer: By `at least` here, does it mean `>=` or `>`? (people could get confused with this so I would either say "at least or equal to" or "strictly larger than" depending on the intent).

Reviewer 2: What about "*abs_tol* is the absolute tolerance; it must be nonnegative."?  Ditto for rel_tol: "*rel_tol* must be nonnegative."

FYI: PEP and docs now specify rel_tol as > 0.  While code checks >= condition instead.  I'm not sure if we should stay with code or with docs.  Nothing harmful IMO in rel_tol=0, especially if abs_tol != 0.

Reviewer: > nonnegative

I usually use that formulation to avoid the overly verbose "greater/less than or equal" so I like it.

> I'm not sure if we should stay with code or with docs

I think it's better to stick with the code because it allows people to put it to 0 if they want. Otherwise they might think "hey, rel_tol = 0 is special and I'm not allowed to do it" (or it may comes from some other computed value). Saying that both `abs_tol` and `rel_tol` must be nonnegative is simpler (and correct) for learning how to use that function.

---

cc @willingc

Reviewer 3: I agree that the doc should be changed to acknowledge rel_tol=0.0.  As noted on both the issue and above, the effect is to make isclose depend only on abs_tol.  This could be mentioned in the rel_tol sentence.  Perhaps this is why the code deviated from the PEP.

Reviewer 3: A change was made.

>    When comparing to ``0.0``, the default tolerance values means that only

Reviewer 4: Hmm... I'm not sure that this wording clarifies the behavior.
```suggestion
   When comparing to ``0.0``, the relative tolerance value means
```

Reviewer 4: I would make a similar change in the other file too.

Reviewer 4: @skirpichev Please re-review. Now that I'm reflecting on this, I'm not sure my suggestion about `relative` is correct. Perhaps it should have been `default absolute`. Thanks!

Author: I think it should be `default absolute`. My opinion may be wrong and is just for reference.

Reviewer 2: > the default tolerance values

The point was that defaults (both for absolute=0.0 and relative tolerances) means that only 0.0 is close to 0.0.  Maybe this: "*abs_tol* is the absolute tolerance; it must be nonnegative.  Only ``0.0` is close to ``0.0`` per default.  Pass an appropriate absolute tolerance to compare with ``0.0``."

Reviewer 4: Thanks @skirpichev and @Xiaokang2022. Let's use @skirpichev's suggested wording. Thanks!

Author: @skirpichev @willingc Please review again. Thanks!

Reviewer 3: I am responding in a fresh comment in which I will request changes.  Please do not merge before that.

   tolerance is ``1e-09``, which assures that the two values are the same
   within about 9 decimal digits.  *rel_tol* must be greater than zero.

>    *abs_tol* is the absolute tolerance; it must be nonnegative.  Only ``0.0``
>    is close to ``0.0`` per default.  Pass an appropriate absolute tolerance to
>    compare with ``0.0``.

Reviewer 3: The initial PR is my suggestion in https://github.com/python/cpython/issues/126165#issuecomment-2448116678 that replaced 2 sentences with 3:
"*abs_tol* is the minimum absolute tolerance; it must be at least ``0.0``.
When comparing to ``0.0``, the default *abs_tol* of ``0.0`` means that only ``0.0`` is close to ``0.0``.
Pass a larger tolerance to allow for floating point inaccuracies."
I respond to several comments elsewhere by discussing changes to each sentence separately and finish with a suggested rewrite of all 3.

Sentence 1: removing 'minimum', carried over from the original, looks fine to me.  'nonnegative' is equivalent to 'at least 0.0'.  In English, the latter means 'equal to or greater than 0.0' ('>= 0.0').  Preference here is partly a matter of taste and judgment of what is better for readers with a non-English first language.

Sentence 2: Since `isclose(x, 0)` is computed as `abs(x) <= rel_tol * abs(x)`; the result is False for x != 0 as long as rel_tol < 1.  I implicitly assumed this since if rel_tol >= 1, The result is True for any x.  The initial replacement of "abs_tol of 0.0" by "tolerance values" acknowledges dependence on a sane rel_tol at the cost of removing the default and being more confusing.  Specifying the default is part of the purpose of the sentence as I wrote it.  I think my original phrase was better.  I also dislike the replacement sentence "Only ``0.0`` is close to ``0.0`` per default." To me, "per default" is an awkward legalism.

Sentence 3: My suggest is a bit awkward.  In the suggested replacement, "Pass an appropriate absolute tolerance to compare with ``0.0``.", 'positive' is missing and 'to compare with 0.0' is redundant as this is the context started in the sentence before.

Here is a new abs_tol entry that takes into account the comments above.  Among other things, it separates statement of the default from discussion of closeness to 0.   I will work best with changes elsewhere that I will propose in my review comment. The markup may need tweaks.
```suggestion
   *abs_tol* is the absolute tolerance; it defaults to ``0.0`` and it must be nonnegative.  When comparing ``x`` to ``0.0``, ``isclose(x, 0)`` is computed as ``abs(x) <= rel_tol  * abs(x)``, which is false for any ``x`` and rel_tol less than ``1.0``.  So add an appropriate positive abs_tol argument to the call. 
```

Reviewer 3: With respect to point 2, I think rel_tol >= 1 is likely a mistake of entering a percentage, such as ``1`` when ``.01`` was meant.  So we might say that it must be nonnegative and normally is much less than 1.

Author: I've made a few slight formatting changes to your suggestion:

```diff
- less than one
+ less than ``1.0``
```

```diff
- which is false for
+ which is ``False`` for
```

Reviewer 3: Looks good.  I will wait a day before merging.
